File: cqueue/c_queue.cc
/*
 * Copyright 2018 Infostellar, Inc.
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street,
 * Boston, MA 02110-1301, USA.
 */

#include "c_queue.h"

c_queue::c_queue(int buffer_size) : queue_(buffer_size) {}

void c_queue::push(const std::string &item) {
  bool pushed = queue_.push(item);
  if (pushed) {
    condition_var_.notify_one();
  }
}

std::string c_queue::pop() {
  std::string a;
  queue_.pop(a);
  return a;
}

std::string c_queue::block_pop() {
  std::string a;
  std::unique_lock<std::mutex> lock(mutex_);
  condition_var_.wait(lock);
  // If woken up spuriously, will return an empty string.
  queue_.pop(a);
  return a;
}

void c_queue::wake() {
  condition_var_.notify_one();
}

uint64_t c_queue::get_ptr() {
  return reinterpret_cast<uint64_t>(this);
}

c_queue *c_queue_from_ptr(uint64_t ptr) {
    return reinterpret_cast<c_queue *>(ptr);
}

File: cqueue/c_queue.go
/*
 * Starcoder - a server to read/write data from/to the stars, written in Go.
 * Copyright (C) 2018 InfoStellar, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cqueue

// Thin wrapper around the string_queue C++ class so
// we only need to deal with unrecognized variables in this file.
type CStringQueue struct {
	queue SwigcptrString_queue
}

func NewCStringQueue(buffer_size int) *CStringQueue {
	if val, ok := NewString_queue(buffer_size).(SwigcptrString_queue); ok {
		return &CStringQueue{
			queue: val,
		}
	}
	return nil
}

func (q *CStringQueue) GetPtr() uint64 {
	return q.queue.Get_ptr()
}

func (q *CStringQueue) Pop() string {
	return q.queue.Pop()
}

func (q *CStringQueue) BlockingPop() string {
	return q.queue.Blocking_pop()
}

func (q *CStringQueue) Close() {
	q.queue.Close()
}

func (q *CStringQueue) Closed() bool {
	return q.queue.Closed()
}

func (q *CStringQueue) Push(str string) {
	q.queue.Push(str)
}

func (q *CStringQueue) Delete() {
	DeleteString_queue(q.queue)
}

func CStringQueueFromPtr(ptr uint64) *CStringQueue {
	if val, ok := String_queueQueue_from_pointer(ptr).(SwigcptrString_queue); ok {
		return &CStringQueue{
			queue: val,
		}
	}
	return nil
}
